File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int preindex=0;
    TreeNode* BuildTree(vector<int>& preorder, vector<int>& inorder, int is,int ie)
    {
        if(is>ie)
            return NULL;
        TreeNode* root = new TreeNode(preorder[preindex++]); 
        int index;
        for(int i=is;i<=ie;i++)
        {
            if(inorder[i]==root->val)
            {
                index=i;
                break;
            }   
        }
        root->left = BuildTree(preorder,inorder,is,index-1);
        root->right = BuildTree(preorder,inorder,index+1,ie);
        return root;
    }
    
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int is=0;
        int ie=preorder.size()-1;
                
        return BuildTree(preorder,inorder,is,ie);
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer,Integer> map = new HashMap<>();
        int n = inorder.length;
        for(int i=0;i<n;i++)
            map.put(inorder[i],i);
        TreeNode root = build(preorder, 0,n-1,inorder, 0,n-1,map);
        return root;
    }
    
    private TreeNode build(int[] preorder, int preStart, int preEnd,int[] inorder, int inStart, int inEnd,HashMap<Integer,Integer> map){
        if(preStart>preEnd || inStart>inEnd) return null;
        
        TreeNode root = new TreeNode(preorder[preStart]);
        
        int index = map.get(root.val);
        int diff = index-inStart;
        
        root.left = build(preorder, preStart+1,preStart+diff , inorder,inStart,index-1,map);
        root.right = build(preorder,preStart+diff+1,preEnd,inorder,index+1,inEnd,map);
        return root;
    }
}